In [1]:
import pandas as pd
import plotly.express as px
import warnings
warnings.filterwarnings('ignore')

# MN Contribution EDA

## 1. Read in and Preprocess Datasets

### 1.1 Read in datasets

In [2]:
# Read in candidate-recipient contribution data
df_ag = pd.read_csv('/project/data/cand_con.csv/AG.csv')
df_ap = pd.read_csv('/project/data/cand_con.csv/AP.csv')
df_dc = pd.read_csv('/project/data/cand_con.csv/DC.csv')
df_gc = pd.read_csv('/project/data/cand_con.csv/GC.csv')
df_house = pd.read_csv('/project/data/cand_con.csv/House.csv')
df_sa = pd.read_csv('/project/data/cand_con.csv/SA.csv')
df_sc = pd.read_csv('/project/data/cand_con.csv/SC.csv')
df_senate = pd.read_csv('/project/data/cand_con.csv/Senate.csv')
df_ss = pd.read_csv('/project/data/cand_con.csv/SS.csv')
df_st = pd.read_csv('/project/data/cand_con.csv/ST.csv')

# Read in non-candidate-recipient contribution data
df_non_cand = pd.read_csv('/project/data/non_candidate_con.csv')

10 datasets on candidate-recipient contributions and 1 dataset on non-candidate-recipient contributions. They are seperate and not relational

### 1.2 Check for DataFrames' column consistency

In [3]:
# First check for DataFrames' column numbers
df_lst = [df_ag, df_ap, df_dc, df_gc, df_house, df_sa, df_sc, df_ss, df_st, df_senate, df_non_cand]
df_lens = []
for df in df_lst:
    df_lens.append(df.shape[1])
print(df_lens)

[13, 13, 13, 13, 13, 13, 13, 13, 13, 13, 11]


In [4]:
from utils.MN_util import datasets_col_consistent

datasets_col_consistent(df_lst[:-1])

All dfs have consistent columns


### 1.3 Adjust for DataFrames' column consistency

In [5]:
df_ag.columns, df_non_cand.columns

(Index(['OfficeSought', 'Party', 'District', 'CandRegNumb', 'CandFirstName',
        'CandLastName', 'CommitteeName', 'DonationDate', 'DonorType',
        'DonorName', 'DonationAmount', 'InKindDonAmount',
        'InKindDescriptionText'],
       dtype='object'),
 Index(['PCFRegNumb', 'Committee', 'ETType', 'ETSubType', 'DonationDate',
        'DonorType', 'DonorRegNumb', 'DonorName', 'DonationAmount',
        'InKindDonAmount', 'InKindDescriptionText'],
       dtype='object'))

Based on the project need and dataset consistency, use these columns: RegNumb, RecipientType, OfficeSought, CandFirstName,  CandLastName, Committee, DonationDate, DonorType, DonorName, DonationAmount, InKindDonAmount, InKindDescriptionText

In [6]:
from utils.MN_util import preprocess_candidate_df

df_ag = preprocess_candidate_df(df_ag)
df_ap = preprocess_candidate_df(df_ap)
df_dc = preprocess_candidate_df(df_dc)
df_gc = preprocess_candidate_df(df_gc)
df_house = preprocess_candidate_df(df_house)
df_sa = preprocess_candidate_df(df_sa)
df_sc = preprocess_candidate_df(df_sc)
df_ss = preprocess_candidate_df(df_ss)
df_st = preprocess_candidate_df(df_st)
df_senate = preprocess_candidate_df(df_senate)

In [7]:
from utils.MN_util import preprocess_noncandidate_df
df_non_cand = preprocess_noncandidate_df(df_non_cand)

In [8]:
from utils.MN_util import preprocess_contribution_df

new_df_lst = [df_ag, df_ap, df_dc, df_gc, df_house, df_sa, df_sc, df_ss, df_st, 
              df_senate, df_non_cand]

contribution_df = preprocess_contribution_df(new_df_lst)

In [9]:
contribution_df['DonorType'].unique()

array(['I', 'F', 'C', 'O', 'L', 'P', 'H', 'U', 'S', nan, 'B'],
      dtype=object)

In [10]:
contribution_df.dtypes

OfficeSought                 object
RegNumb                       int64
CandFirstName                object
CandLastName                 object
Committee                    object
Date                 datetime64[ns]
DonorType                    object
DonorName                    object
Amount                      float64
InKindAmount                float64
InKindDescription            object
RecipientType                object
Year                          int64
TotalAmount                 float64
dtype: object

#### Donor Types:
1. C: Candidate Committee (limited to state-level candidates who had a principal campaign committee registered with the Board from which the contribution was made)
2. I: Non-lobbyist individual 
3. L: Lobbyist  
4. F: Political Committee/Fund  
5. S: Supporting association of a political fund registered with the Board that donates to its own political fund
6. P: Political party unit
7. H: Local candidate committee (limited to candidates within Hennepin County who satisfy the definition of local candidate, did not exist until 2022)
8. O: Other (catch-all category that in some cases includes businesses, supporting associations of political funds registered with the Board that donate to their own political fund, associations that are not registered with the Board, and any entity that does not fall within one of the other categories)
9. U: Association not registered with the Board (may include a committee registered with the FEC or a regulatory committee in another state, a 501(c)(4), 501(c)(6), or 527 nonprofit organization, the campaign committee of a candidate for local office (excluding certain Hennepin County candidates from 2022 onward), etc.)
10. B: Business (company & corporation)

In [11]:
contribution_df['RecipientType'].unique()

array(['Candidate', 'PCF', 'PTU'], dtype=object)

In [12]:
contribution_df['OfficeSought'].unique()

array(['AG', nan, 'GC', 'House', 'Senate', 'SA', 'SS', 'SC', 'DC', 'AP',
       'ST'], dtype=object)

#### Recipient Types:
- Candidate
- PCF: Political committee or fund
- PTU: Political party unit

#### Office Types (within candidate recipient):
- AG = Attorney General
- AP = State Appeals Court Judge
- DC = State District Court Judge
- GC = Governor
- House = State Representative
- SA = State Auditor
- SC = State Supreme Court Justice
- SS = Secretary of State
- ST = State Treasurer (this office was abolished in 2003 and no longer exists)
- Senate = State Senator

### 1.4 Check Missing Values

In [13]:
contribution_df.isna().sum()

OfficeSought          483861
RegNumb                    0
CandFirstName         483861
CandLastName          483861
Committee                467
Date                     536
DonorType                580
DonorName                477
Amount                     0
InKindAmount               0
InKindDescription    3508947
RecipientType              0
Year                       0
TotalAmount                0
dtype: int64

In [14]:
print('Total number of contribution entries = ', len(contribution_df))
no_amount = len(contribution_df[contribution_df['TotalAmount'] == 0])
print('Total number of nonclassifiable contribution amount = ', no_amount)
no_don = contribution_df['DonorName'].isna().sum()
print('Total number of nonclassifiable donors = ', no_don)
unclassifiable_prop = round((no_amount+no_don)/len(contribution_df),6)
print('Proportion of nonclassifiable entries =', f"{unclassifiable_prop*100}%")

Total number of contribution entries =  3548873
Total number of nonclassifiable contribution amount =  336051
Total number of nonclassifiable donors =  477
Proportion of nonclassifiable entries = 9.4827%


1. 467 of the 'nan' contributions belong to "Registration Fee for Netroots Event", which is a non-profit organization that help progressive activists. Based on their column values, these contributions have no monetary amount, no donor, and no recipient.
2. 483861 nan OfficeSought are mostly those recipients are non-candidates
3. Contribution entries with no contribution amount, recipient information, or donor information should be dropped

### 1.5 Drop Non-classifiable Contribution Data

In [15]:
from utils.MN_util import drop_nonclassifiable
contribution_df = drop_nonclassifiable(contribution_df)

: 

## 2. Top 10
### 2.1 Top 10 Donors

In [ ]:
filtered_df = contribution_df[(contribution_df['Year'] >= 2018) & (contribution_df['Year'] <= 2023) & (contribution_df['Year'] != -1)]
donation_by_year = filtered_df.groupby('Year')['TotalAmount'].sum()
donation_by_year

Year
2018    47781848.87
2019    25764905.60
2020    34614419.07
2021    26898778.21
2022    57692246.12
2023      264961.92
Name: TotalAmount, dtype: float64

In [ ]:
# Group by 'Year' and 'Contributor' for total contribution for each contributor in each year
don_by_year_contributor = filtered_df.groupby(
    ['Year', 'DonorName'])['TotalAmount'].sum().reset_index()

# Find the top 10 contributors
top_10_contributors = don_by_year_contributor.groupby('Year').apply(
    lambda group: group.nlargest(10, 'TotalAmount')).reset_index(drop=True)

In [ ]:
top_10_contributors

Year                                        DonorName  TotalAmount
0   2018                 Minn DFL State Central Committee    519799.42
1   2018           Minn Chamber of Commerce Leadership Fd    462600.00
2   2018              Minn Nurses Assn Pol Comm  (MNA-PC)    322750.00
3   2018                          MTA PAC (fka SITCO PAC)    301400.00
4   2018                               Education Minn PAC    291200.00
5   2018                       Shakopee Mdewakanton Sioux    290550.00
6   2018          Faegre Baker Daniels State-Reg Pol Fund    276350.00
7   2018                      IBEW Minn State Council PAC    188500.00
8   2018                                  Haselow, Robert    175025.00
9   2018           Minn Dental Political Action Committee    167350.00
10  2019                       Shakopee Mdewakanton Sioux    683300.00
11  2019           Minn Chamber of Commerce Leadership Fd    466000.00
12  2019                    Minn Business Partnership PAC    369750.00
13  2019          Faegre Baker Daniels State-Reg Pol Fund    297650.00
14  2019             Beer PAC-Minn Beer Wholesalers Assoc    279250.00
15  2019           Winthrop & Weinstine PA Political Fund    262750.00
16  2019               Minn CPAs Public Affairs Committee    262250.00
17  2019                      IBEW Minn State Council PAC    257000.00
18  2019                                  Haselow, Robert    237800.00
19  2019             Rural Electric Political Action Comm    236200.00
20  2020                               Education Minn PAC    367950.00
21  2020                          MTA PAC (fka SITCO PAC)    335800.00
22  2020              Minn Nurses Assn Pol Comm  (MNA-PC)    310360.00
23  2020  Laborers District Council of Minn & ND Pol Fund    223750.00
24  2020                                 Haselow, Justine    221800.00
25  2020                                Mendoza, Salvador    210500.00
26  2020              Prairie Island Indian Community PAC    206300.00
27  2020                                  Haselow, Robert    192150.00
28  2020              Faegre Drinker State Political Fund    184150.00
29  2020                    Minn Business Partnership PAC    175000.00
30  2021              Faegre Drinker State Political Fund    470400.00
31  2021           Minn Chamber of Commerce Leadership Fd    349500.00
32  2021                                     Einess, Ward    268200.00
33  2021              North Central States Carpenters PAC    252550.00
34  2021               Minn CPAs Public Affairs Committee    231000.00
35  2021             Rural Electric Political Action Comm    221150.00
36  2021                    Minn Business Partnership PAC    201475.00
37  2021                          Christopherson, Kirsten    173200.00
38  2021             Beer PAC-Minn Beer Wholesalers Assoc    165750.00
39  2021           Winthrop & Weinstine PA Political Fund    159750.00
40  2022                       Shakopee Mdewakanton Sioux    340900.00
41  2022           Minn Chamber of Commerce Leadership Fd    272750.00
42  2022  Laborers District Council of Minn & ND Pol Fund    235050.00
43  2022              Faegre Drinker State Political Fund    202500.00
44  2022                               Education Minn PAC    196258.16
45  2022                          MTA PAC (fka SITCO PAC)    192250.00
46  2022                                  Haselow, Robert    167300.00
47  2022                                Mendoza, Salvador    167000.00
48  2022                                     Mendoza, Mia    154000.00
49  2022              North Central States Carpenters PAC    151500.00
50  2023                                   IBEW Local 292     11400.11
51  2023                                    SEIU local 26      5194.45
52  2023                              Teamsters Local 346      3776.37
53  2023                                    Johnson, Erin      2899.00
54  2023                              Teamsters Local 792      2897.77
55  2023                    

### 2.2 Top 10 Recipients

In [ ]:
contribution_df1 = filtered_df.copy(deep=True)
contribution_df1['CandLastName'].fillna('NA', inplace=True)
contribution_df1['CandFirstName'].fillna('NA', inplace=True)
contribution_df1['Committee'].fillna('NA', inplace=True)
by_year_recipients = contribution_df1.groupby(
    ['Year', 'RegNumb', 'RecipientType', 'CandLastName', 'CandFirstName', 'Committee'])['TotalAmount'].sum().reset_index()

top_10_recipient = by_year_recipients.groupby('Year').apply(
    lambda group: group.nlargest(10, 'TotalAmount')).reset_index(drop=True)

In [ ]:
top_10_recipient

Year  RegNumb RecipientType CandLastName CandFirstName  \
0   2018    18135     Candidate         Walz           Tim   
1   2018    17641     Candidate      Johnson          Jeff   
2   2018    18336     Candidate      Ellison         Keith   
3   2018    18133     Candidate      Wardlow          Doug   
4   2018    18125     Candidate       Murphy          Erin   
5   2018    17653     Candidate        Simon         Steve   
6   2018    18292     Candidate     Pawlenty           Tim   
7   2018    15677     Candidate      Hortman       Melissa   
8   2018    17902     Candidate      Edelson       Heather   
9   2018    12604     Candidate       Davids       Gregory   
10  2019    18135     Candidate         Walz           Tim   
11  2019    11829     Candidate        Cohen       Richard   
12  2019    16964     Candidate      Housley         Karin   
13  2019    15677     Candidate      Hortman       Melissa   
14  2019    12604     Candidate       Davids       Gregory   
15  2019    15317     Candidate    Tomassoni         David   
16  2019    13262     Candidate       Limmer        Warren   
17  2019    18336     Candidate      Ellison         Keith   
18  2019    17040     Candidate        Osmek         David   
19  2019    17653     Candidate        Simon         Steve   
20  2020    18135     Candidate         Walz           Tim   
21  2020    17105     Candidate       Nelson         Carla   
22  2020    17481     Candidate         Hawj         Foung   
23  2020    15501     Candidate        Rosen         Julie   
24  2020    17443     Candidate         Kent         Susan   
25  2020    15667     Candidate       Dibble       D Scott   
26  2020    18021     Candidate      Westlin        Bonnie   
27  2020    15705     Candidate       Senjem         David   
28  2020    17653     Candidate        Simon         Steve   
29  2020    16891     Candidate       Miller        Jeremy   
30  2021    18135     Candidate         Walz           Tim   
31  2021    17653     Candidate        Simon         Steve   
32  2021    18690     Candidate       Jensen         Scott   
33  2021    18336     Candidate      Ellison         Keith   
34  2021    15317     Candidate    Tomassoni         David   
35  2021    15677     Candidate      Hortman       Melissa   
36  2021    16964     Candidate      Housley         Karin   
37  2021    17140     Candidate         Bakk  Thomas (Tom)   
38  2021    12604     Candidate       Davids       Gregory   
39  2021    18133     Candidate      Wardlow          Doug   
40  2022    18135     Candidate         Walz           Tim   
41  2022    17653     Candidate        Simon         Steve   
42  2022    18690     Candidate       Jensen         Scott   
43  2022    18336     Candidate      Ellison         Keith   
44  2022    15677     Candidate      Hortman       Melissa   
45  2022    18731     Candidate      Schultz         James   
46  2022    18133     Candidate      Wardlow          Doug   
47  2022    13262     Candidate       Limmer        Warren   
48  2022    10601     Candidate       Murphy          Mary   
49  2022    18218     Candidate        Blaha         Julie   
50  2023    41291           PCF           NA            NA   
51  2023    18690     Candidate       Jensen         Scott   
52  2023    30689           PCF           NA            NA   
53  2023    41345           PCF           NA            NA   
54  2023    30726           PCF           NA            NA   
55  2023    30013           PCF           NA            NA   
56  2023    30345           PCF           NA            NA   
57  2023    41281           PCF           NA            NA   
58  2023    41262           PCF           NA            NA   
59  2023    30119           PCF           NA            NA   

                                            Committee  TotalAmount  
0                               Tim Walz for Governor   8095546.97  
1                         Johnson (Jeff) for Governor   4060066.88  
2             

## 3. Compare donation by donor and recipient types

### 3.1 Compare donation by donor types

In [ ]:
donor_type_mapping = {
    'B': 'Business',
    'C': 'Candidate committee',
    'F': 'Political committee or fund',
    'H': 'Local candidate committee registered with Hennepin County',
    'I': 'Non-lobbyist individual',
    'L': 'Lobbyist',
    'O': 'Other',
    'P': 'Political party unit',
    'S': 'Self',
    'U': 'Association not registered with the Board'
}

In [ ]:
grouped2 = filtered_df.groupby(['Year', 'DonorType'])['TotalAmount'].sum().reset_index()

grouped2['FullDonorType'] = grouped2['DonorType'].map(donor_type_mapping)

fig = px.bar(
    grouped2,
    x='Year',
    y='TotalAmount',
    color='FullDonorType',
    title='Donations by Donor Type From 2018 To 2023',
    labels={"Year": "Year", "TotalAmount": "Total Contributions", "FullDonorType": "Donor Type"},
    category_orders={"FullDonorType": sorted(donor_type_mapping.values())}
)

fig.show()

In [ ]:
filtered_df2 = contribution_df[contribution_df['Year'] == 2022]
grouped3 = filtered_df2.groupby(['Year', 'DonorType'])['TotalAmount'].sum().reset_index()

grouped3['FullDonorType'] = grouped3['DonorType'].map(donor_type_mapping)

fig = px.bar(
    grouped3,
    x='Year',
    y='TotalAmount',
    color='FullDonorType',
    title='Donations by Donor Type in 2022',
    labels={"DonationYear": "Year", "TotalAmount": "Total Contributions", "FullDonorType": "Donor Type"},
    category_orders={"FullDonorType": sorted(donor_type_mapping.values())}
)

fig.show()

#### Observations and Interpretations
1. Individuals, excluding lobbyists, constitute the largest share of contributions in the MN dataset.
2. The second most substantial contributor category is General Purpose Political Committee or Fund, followed by lobbyists.
3. Contributions from other donor types are notably lower throughout the years.
4. Analyzing a sample from 2018 to 2022, we observe a cyclical pattern with a major increase in contributions, followed by three years of reduced contribution totals. This cycle aligns with the four-year election cycle.
5. From 1998 to 2023, there are several years with significantly lower contribution amount: 1999, 2001, 2003, 2007, 2011.

### 3.2 Compare donation by recipient types

In [ ]:
grouped4 = filtered_df.groupby(['Year', 'RecipientType'])['TotalAmount'].sum().reset_index()

fig = px.bar(
    grouped4,
    x='Year',
    y='TotalAmount',
    color='RecipientType',
    title='Donations by Recipient Type from 2018 to 2023',
    labels={"Year": "Year", "TotalAmount": "Total Contributions", 'RecipientType': 'Recipient Type'},
)

fig.show()

In [ ]:
grouped4 = filtered_df2.groupby(['Year', 'RecipientType'])['TotalAmount'].sum().reset_index()

fig = px.bar(
    grouped4,
    x='Year',
    y='TotalAmount',
    color='RecipientType',
    title='Donations by Recipient Type in 2022',
    labels={"Year": "Year", "TotalAmount": "Total Contributions", 'RecipientType': 'Recipient Type'},
)

fig.show()

#### Observations and Interpretations
1. Candidates, as the recipients, make up the overwhelming majority of contributions.
2. Examining the period from 1998 to 2023, a distinct cyclical pattern emerges, characterized by alternating years of increased and decreased contributions, which may correspond to congressional elections or MN state house representatives elections which take place every two years.
3. Starting in 2012, recipient types "Political Committee or Fund" and "Political Party Unit" began receiving a larger share of contributions compared to prior years.

In [ ]:
race_type_mapping = {
    'AG': 'Attorney General',
    'AP': 'Political committee or fund',
    'DC': 'Candidate committee',
    'GC': 'Other',
    'House': 'Lobbyist',
    'SA': 'State Auditor',
    'SC': 'Local candidate committee registered with Hennepin County',
    'SS': 'Secretary of State',
    'Senate': 'Senate',
}

In [ ]:
grouped5 = filtered_df.groupby(['Year', 'OfficeSought'])['TotalAmount'].sum().reset_index()
grouped5['FullRaceType'] = grouped5['OfficeSought'].map(race_type_mapping)

fig = px.bar(
    grouped5,
    x='Year',
    y='TotalAmount',
    color='FullRaceType',
    title='Donations by Candidate Recipient Race From 2018 To 2023',
    labels={"Year": "Year", "TotalAmount": "Total Contributions", 'FullRaceType': 'Full Race Type'},
    category_orders={"FullRaceType": sorted(race_type_mapping.values())}
)

fig.show()

Clearly, state senators and house representatives recieve the most contributions

In [ ]:
contribution_df[contribution_df['RegNumb']==30119]

OfficeSought  RegNumb CandFirstName CandLastName  \
22               NaN    30119           NaN          NaN   
58               NaN    30119           NaN          NaN   
95               NaN    30119           NaN          NaN   
103              NaN    30119           NaN          NaN   
108              NaN    30119           NaN          NaN   
...              ...      ...           ...          ...   
1621101          NaN    30119           NaN          NaN   
1624800          NaN    30119           NaN          NaN   
2297031          NaN    30119           NaN          NaN   
2298939          NaN    30119           NaN          NaN   
2300737          NaN    30119           NaN          NaN   

                                       Committee       Date DonorType  \
22       IBEW Local 292 Political Education Fund 2023-03-05         O   
58       IBEW Local 292 Political Education Fund 2023-03-09         O   
95       IBEW Local 292 Political Education Fund 2023-02-17         O   
103      IBEW Local 292 Political Education Fund 2023-02-23         O   
108      IBEW Local 292 Political Education Fund 2023-02-28         O   
...                                          ...        ...       ...   
1621101  IBEW Local 292 Political Education Fund 2012-10-18         O   
1624800  IBEW Local 292 Political Education Fund 2012-10-11         O   
2297031  IBEW Local 292 Political Education Fund 2008-12-30         O   
2298939  IBEW Local 292 Political Education Fund 2008-10-14         O   
2300737  IBEW Local 292 Political Education Fund 2008-10-22         O   

              DonorName  Amount  InKindAmount InKindDescription RecipientType  \
22       IBEW Local 292   55.76           0.0               NaN           PCF   
58       IBEW Local 292  262.22           0.0               NaN           PCF   
95       IBEW Local 292  550.42           0.0               NaN           PCF   
103      IBEW Local 292  133.66           0.0               NaN           PCF   
108      IBEW Local 292   75.10           0.0               NaN           PCF   
...                 ...     ...           ...               ...           ...   
1621101  IBEW Local 292   62.30           0.0               NaN           PCF   
1624800  IBEW Local 292  196.12           0.0               NaN           PCF   
2297031  IBEW Local 292   59.31           0.0               NaN           PCF   
2298939  IBEW Local 292  351.60           0.0               NaN           PCF   
2300737  IBEW Local 292  112.14           0.0               NaN           PCF   

         Year  TotalAmount  
22       2023        55.76  
58       2023       262.22  
95       2023       550.42  
103      2023       133.66  
108      2023        75.10  
...       ...          ...  
1621101  2012        62.30  
1624800  2012       196.12  
2297031  2008        59.31  
2298939  2008       351.60  
2300737  2008       112.14  

[1010 rows x 14 columns]

# MN Expenditure EDA

### 1. Read in and Preprocess Datasets

In [2]:
df_general = pd.read_csv('/project/data/general_exp_con.csv')
df_independent = pd.read_csv('/project/data/independent_exp.csv')

In [3]:
from utils.MN_util import preprocess_general_exp_df
df_general = preprocess_general_exp_df(df_general)

In [4]:
from utils.MN_util import preprocess_independent_exp_df
df_independent = preprocess_independent_exp_df(df_independent)

In [5]:
df_independent.columns==df_general.columns

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True])

In [6]:
df_independent.dtypes

SpenderRegNum                int64
SpenderName                 object
SpenderType                 object
VendorName                  object
Amount                     float64
UnpaidAmount               float64
Date                        object
Year                         int64
Purpose                     object
Type                        object
In-kind?                    object
InKindDescription           object
AffectedCommitteeName       object
AffectedCommitteeRegNum    float64
dtype: object

In [7]:
from utils.MN_util import preprocess_expenditure_df

df_expenditure = preprocess_expenditure_df([df_general, df_independent])

In [8]:
df_expenditure

SpenderRegNum                                        SpenderName  \
0               15863                 Hooten, Carol Dist Court Committee   
1               16008                   Faust, Timothy D House Committee   
2               16248                   Swails, Marsha G House Committee   
3               16248                   Swails, Marsha G House Committee   
4               16276                     Sawatzke, Pat Senate Committee   
...               ...                                                ...   
664185          80026  MN Assoc of Professional Employees Political Fund   
664186          80026  MN Assoc of Professional Employees Political Fund   
664187          80026  MN Assoc of Professional Employees Political Fund   
664188          80026  MN Assoc of Professional Employees Political Fund   
664189          80026  MN Assoc of Professional Employees Political Fund   

       SpenderType                                VendorName   Amount  \
0              PCC                                       NaN   102.84   
1              PCC                      Kanabec Publications   429.11   
2              PCC                                       NaN  1293.77   
3              PCC     Women Candidate Development Coalition  1293.76   
4              PCC                            UND Foundation   500.00   
...            ...                                       ...      ...   
664185         PCF  MN Association of Professional Employees   304.28   
664186         PCF                         No Coast Workshop   217.39   
664187         PCF                         No Coast Workshop   217.40   
664188         PCF                         No Coast Workshop   144.32   
664189         PCF                         No Coast Workshop   144.33   

        UnpaidAmount        Date  Year  \
0                0.0  10/10/2015  2015   
1                0.0  02/18/2015  2015   
2                0.0  03/25/2015  2015   
3                0.0  03/25/2015  2015   
4                0.0  07/30/2015  2015   
...              ...         ...   ...   
664185           0.0  10/21/2020  2020   
664186           0.0  10/13/2016  2016   
664187           0.0  10/13/2016  2016   
664188           0.0  11/03/2016  2016   
664189           0.0  11/03/2016  2016   

                                                  Purpose  \
0                               Repayment of part of loan   
1       Advertising - Print: $4.96 was service charge ...   
2                                                     NaN   
3                              contribution to non-profit   
4                                 charitable contribution   
...                                                   ...   
664185                      Employee Expense: Staff costs   
664186  Printing and Photocopying: Independent Expendi...   
664187  Printing and Photocopying: Independent Expendi...   
664188  Postage/ Delivery: Independent Expenditure Mai...   
664189  Postage/ Delivery: Independent Expenditure Mai...   

                           Type In-kind? InKindDescription  \
0            Other Disbursement       No               NaN   
1          Campaign Expenditure       No               NaN   
2                  Contribution       No               NaN   
3            Other Disbursement       No               NaN   
4            Other Disbursement       No               NaN   
...                         ...      ...               ...   
664185  Independent Expenditure       No               NaN   
664186  Independent Expenditure       No               NaN   
664187  Independent Expenditure       No               NaN   
664188  Independent Expenditure       No               NaN   
664189  Independent Expenditure       No               NaN   

                          AffectedCommitteeName  AffectedCommitteeRegNum  
0                                           NaN                      NaN  
1                                           NaN                      NaN  
2         

In [9]:
print(df_expenditure['SpenderType'].unique())
print(df_expenditure['Type'].unique())

['PCC' 'PCF' 'PTU']
['Other Disbursement' 'Campaign Expenditure' 'Contribution'
 'Non-Campaign Disbursement' 'General Expenditure'
 'Ballot Question Expenditure' 'Party Unit' 'Candidate Committee'
 'Political Committee/Fund' 'Independent Expenditure']


### 2. Check Missing Values and Drop Unclassifiable Data

In [10]:
df_expenditure.isna().sum()

SpenderRegNum                   0
SpenderName                    81
SpenderType                     0
VendorName                  38386
Amount                          0
UnpaidAmount                    0
Date                            0
Year                            0
Purpose                     66290
Type                            0
In-kind?                        0
InKindDescription          639990
AffectedCommitteeName      236995
AffectedCommitteeRegNum    236693
dtype: int64

In [11]:
print('Total number of expenditure entries =', len(df_expenditure))
no_amount = len(df_expenditure[df_expenditure['Amount'] == 0])
print('Total number of nonclassifiable expenditure amount =', no_amount)
no_spender = df_expenditure['SpenderName'].isna().sum()
print('Total number of nonclassifiable spenders =', no_spender)

unclassifiable_prop = round((no_amount+no_spender)/len(df_expenditure),6)
print('Proportion of nonclassifiable entries =', f"{unclassifiable_prop*100}%")

Total number of expenditure entries = 664190
Total number of nonclassifiable expenditure amount = 355
Total number of nonclassifiable spenders = 81
Proportion of nonclassifiable entries = 0.0656%


In [12]:
from utils.MN_util import drop_nonclassifiable_exp
df_expenditure = drop_nonclassifiable_exp(df_expenditure)

In [13]:
filtered_df = df_expenditure[(df_expenditure['Year'] >= 2018) & (df_expenditure['Year'] <= 2023)]
expenditure_by_year = filtered_df.groupby('Year') 
expenditure_by_year['Amount'].sum()

Year
2018    1.379913e+08
2019    2.612436e+07
2020    1.113496e+08
2021    2.680744e+07
2022    2.799756e+09
2023    4.155524e+06
Name: Amount, dtype: float64

### 3. Top Spenders and Vendors (2018-2023)

#### 3.1 Top 10 Spenders

In [14]:
exp_by_year_spender = filtered_df.groupby(
    ['Year', 'SpenderName'])['Amount'].sum().reset_index()

top_10_spender = exp_by_year_spender.groupby('Year').apply(
    lambda group: group.nlargest(10, 'Amount')).reset_index(drop=True)

In [15]:
top_10_spender

Year                                        SpenderName        Amount
0   2018                     MN DFL State Central Committee  1.634667e+07
1   2018        Alliance for a Better Minnesota Action Fund  9.388784e+06
2   2018                                   DGA Victory Fund  7.639284e+06
3   2018                                          2024 Fund  5.702886e+06
4   2018                                   DFL House Caucus  5.409447e+06
5   2018                                 Education Minn PAC  5.224434e+06
6   2018                            Walz, Tim Gov Committee  4.608298e+06
7   2018                                     MN Victory PAC  3.785053e+06
8   2018                WIN Minnesota Political Action Fund  3.263076e+06
9   2018                                               HRCC  3.079447e+06
10  2019                     MN DFL State Central Committee  3.582806e+06
11  2019                                   DGA Victory Fund  2.554925e+06
12  2019                Deloitte Political Action Committee  2.316457e+06
13  2019                                 Education Minn PAC  1.066196e+06
14  2019                                  DFL Senate Caucus  8.271695e+05
15  2019                           Republican Party of Minn  7.829564e+05
16  2019                                   DFL House Caucus  6.504046e+05
17  2019                                               HRCC  6.390662e+05
18  2019                          Senate Victory Fund (SVF)  6.190183e+05
19  2019                                  DLCC Victory Fund  4.963511e+05
20  2020                     MN DFL State Central Committee  1.186727e+07
21  2020        Alliance for a Better Minnesota Action Fund  7.422896e+06
22  2020                                  DFL Senate Caucus  7.156607e+06
23  2020                                   DFL House Caucus  6.292336e+06
24  2020                WIN Minnesota Political Action Fund  5.211757e+06
25  2020                                          2024 Fund  4.862000e+06
26  2020                          Senate Victory Fund (SVF)  4.144662e+06
27  2020                                 Education Minn PAC  3.919966e+06
28  2020             SEIU Minn State Council Political Fund  2.728316e+06
29  2020                                   DGA Victory Fund  2.432946e+06
30  2021                     MN DFL State Central Committee  4.348914e+06
31  2021                                  Plan for Progress  1.382791e+06
32  2021                                 Education Minn PAC  1.360038e+06
33  2021                            Walz, Tim Gov Committee  1.324421e+06
34  2021                Deloitte Political Action Committee  1.284125e+06
35  2021                                  DFL Senate Caucus  6.201821e+05
36  2021              MN State Council of UNITE HERE Unions  6.034500e+05
37  2021                           Republican Party of Minn  5.811059e+05
38  2021                                   DFL House Caucus  5.769969e+05
39  2021                                   DGA Victory Fund  5.041542e+05
40  2022                     MN DFL State Central Committee  5.299254e+08
41  2022        Alliance for a Better Minnesota Action Fund  4.077091e+08
42  2022                                   DFL House Caucus  2.187665e+08
43  2022                                     MN for Freedom  1.541938e+08
44  2022           Safe Accessible Fair Elections Minnesota  1.314013e+08
45  2022                    DAGA MN People's Lawyer Project  1.109658e+08
46  2022  Advance Minnesota Independent Expenditure Comm...  9.911244e+07
47  2022                                               HRCC  9.192176e+07
48  2022                 MN Jobs Coalition Legislative Fund  9.078141e+07
49  2022                                      iVote Fund MN  7.980807e+07
50  2023                                        All of Mpls  6.872487e+05
51  2023    Laborers District Council of Minn & ND Pol Fund  4.739393e+05
52  2023                North Central States Carpenters PAC  4.647250e+05
53  2023

#### 3.2 Top 10 Vendors

In [16]:
exp_by_year_vendor = filtered_df.groupby(
    ['Year', 'VendorName'])['Amount'].sum().reset_index()

top_10_vendor = exp_by_year_vendor.groupby('Year').apply(
    lambda group: group.nlargest(10, 'Amount')).reset_index(drop=True)

In [17]:
top_10_vendor

Year                                       VendorName        Amount
0   2018                             Great American Media  7.092650e+06
1   2018                 Democratic Governors Association  5.096131e+06
2   2018                                   Clarify Agency  4.289870e+06
3   2018                                       Nebo Media  3.538082e+06
4   2018                               Drake Bank Federal  3.040606e+06
5   2018                                         GMMB Inc  2.512423e+06
6   2018                                 Berlin Rosen LTD  1.446830e+06
7   2018                                         Ax Media  1.370301e+06
8   2018                    Strategic Media for Placement  1.326959e+06
9   2018                              Education Minnesota  1.302605e+06
10  2019   Democratic Governors Association - Mississippi  2.434730e+06
11  2019                               Drake Bank Federal  1.060554e+06
12  2019  Republican Party of Minnesota - Federal Account  4.960728e+05
13  2019                             Canal Partners Media  3.626000e+05
14  2019                            SEIU MN State Council  3.363778e+05
15  2019                                     Western Bank  3.161543e+05
16  2019                                  Pinnacle Direct  2.952088e+05
17  2019                           Seven Corners Printing  2.412480e+05
18  2019                            EFTPS Fed Tax Payment  1.835754e+05
19  2019                                 Perkins Coie LLP  1.743259e+05
20  2020                                   Clarify Agency  5.621665e+06
21  2020                               Drake Bank Federal  2.728263e+06
22  2020                 Democratic Governors Association  2.397500e+06
23  2020                                        Left Hook  2.380277e+06
24  2020                        WIN Minnesota Federal PAC  2.220500e+06
25  2020                                  FP1 Digital LLC  2.092333e+06
26  2020                             Canal Partners Media  1.448225e+06
27  2020                   Sage Media Planning  Placement  1.375934e+06
28  2020                                Gumbinner  Davies  1.272660e+06
29  2020                              Education Minnesota  1.008506e+06
30  2021                               Drake Bank Federal  1.539032e+06
31  2021                                        AllofMpls  1.229037e+06
32  2021                              Education Minnesota  9.120825e+05
33  2021                            Robin for Minneapolis  6.014500e+05
34  2021                             Cleansweep Campaigns  4.226354e+05
35  2021  Republican Party of Minnesota - Federal Account  3.682714e+05
36  2021                            EFTPS Fed Tax Payment  2.897723e+05
37  2021                               McCool Digital LLC  2.571873e+05
38  2021                                Yes 4 Minneapolis  2.190000e+05
39  2021      Democratic Attorneys General Assocation Inc  2.136878e+05
40  2022                                   Clarify Agency  2.647481e+08
41  2022                             Great American Media  2.582326e+08
42  2022                             Canal Partners Media  1.593249e+08
43  2022                                 Berlin Rosen LTD  1.105615e+08
44  2022                                  Red Eagle Media  1.082093e+08
45  2022                               Deliver Strategies  1.074506e+08
46  2022                                       Nebo Media  8.849665e+07
47  2022                   Sage Media Planning  Placement  8.776352e+07
48  2022                                  FP1 Digital LLC  8.138634e+07
49  2022                Schultz, Jennifer House Committee  8.054502e+07
50  2023                   MN DFL State Central Committee  7.828450e+05
51  2023                                     GRSG Company  5.000000e+05
52  2023                             The People's Canvass  4.230976e+05
53  2023                             MN DFL Senate Caucus  2.496439e+05
54  2023                              MN DFL

### 4. Compare Expenditure by Spender Type and Vendor Type (2018-2023)

In [18]:
spendor_type_mapping = {
    'PCC': 'Candidate', 
    'PCF': 'Political Committee and Fund', 
    'PTU': 'Political Party Unit'
}

grouped = filtered_df.groupby(['Year', 'SpenderType'])['Amount'].sum().reset_index()
grouped['FullSpenderType'] = grouped['SpenderType'].map(spendor_type_mapping)

In [21]:
fig = px.bar(
    grouped,
    x='Year',
    y='Amount',
    color='FullSpenderType',
    title='Expenditure by Spender Type From 2018 To 2023',
    labels={"Year": "Year", "TotalAmount": "Total Expenditure Log", "FullSpenderType": "Spender Type"},
    category_orders={"FullSpenderType": sorted(spendor_type_mapping.values())}
)
fig.update_yaxes(type='log')
fig.show()